core/trino-main/src/main/java/io/trino/util/SystemUtils.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.util;
+
+import com.google.common.base.StandardSystemProperty;
+import oshi.SystemInfo;
+import oshi.hardware.CentralProcessor;
+
+public final class SystemUtils
+{
+    private static final CentralProcessor PROCESSOR = new SystemInfo().getHardware().getProcessor();
+
+    private SystemUtils() {}
+
+    public static int getAvailablePhysicalProcessorCount()
+    {
+        String osArch = StandardSystemProperty.OS_ARCH.value();
+        // logical core count (including container cpu quota if there is any)
+        int availableProcessorCount = Runtime.getRuntime().availableProcessors();
+        int totalPhysicalProcessorCount;
+        if ("amd64".equals(osArch) || "x86_64".equals(osArch)) {
+            // Oshi can recognize physical processor count (without hyper threading) for x86 platforms.
+            // However, it doesn't correctly recognize physical processor count for ARM platforms.
```

Comment:
what does it do on ARM?

Reply:
src/test/resources/sql/presto/tpch/q07.plan.txt

```
Yes, CI is actively conducted on Linux ARM hardware and other platforms will be added when hardware
is available for such testing. Note that many features (e.g., CPUID, and processor identification such as
 family, model, stepping, and vendor frequency) are based on Intel chips and may have
different corresponding meanings.
```

Reply:
Does the method throw?

Reply:
> Does the method throw?

no, see https://github.com/trinodb/trino/pull/10088#issue-1064670692
```
r5.8xlarge (graviton)

no docker:
OSHI physical processors: 1
OSHI logical processors: 32
Available processors: 32
Arch: aarch64

docker:
OSHI physical processors: 1
OSHI logical processors: 32
Available processors: 32
Arch: aarch6

docker (cpu limit=2):
OSHI physical processors: 1
OSHI logical processors: 32
Available processors: 2
Arch: aarch64
```

I don't think it can distinguish (or if it's possible to) physical from logical from `/proc/cpuinfo`